File: TaskCLI/cpp/src/model/task.cpp
// src/model/task.cpp

#include "model/task.h"
#include <algorithm>

using namespace std::chrono;

Task::Task(uint32_t id, const std::string& title, uint32_t user_id)
    : id(id), title(title), user_id(user_id) {
    priority = Priority::Medium;
    status = Status::Todo;
    created_at = system_clock::now();
    updated_at = system_clock::now();
}

void Task::set_description(const std::string& desc) {
    description = desc;
    updated_at = system_clock::now();
}

void Task::set_priority(Priority p) {
    priority = p;
    updated_at = system_clock::now();
}

void Task::set_tags(const std::vector<std::string>& t) {
    tags = t;
    updated_at = system_clock::now();
}

void Task::update_status(Status s) {
    status = s;
    updated_at = system_clock::now();
}

void Task::update_priority(Priority p) {
    priority = p;
    updated_at = system_clock::now();
}

void Task::update_title(const std::string& new_title) {
    title = new_title;
    updated_at = system_clock::now();
}

bool Task::matches(const std::string& keyword) const {
    std::string lower_keyword = keyword;
    std::transform(lower_keyword.begin(), lower_keyword.end(), lower_keyword.begin(), ::tolower);
    
    // 检查标题
    std::string lower_title = title;
    std::transform(lower_title.begin(), lower_title.end(), lower_title.begin(), ::tolower);
    if (lower_title.find(lower_keyword) != std::string::npos) return true;
    
    // 检查描述
    if (description.has_value()) {
        std::string lower_desc = *description;
        std::transform(lower_desc.begin(), lower_desc.end(), lower_desc.begin(), ::tolower);
        if (lower_desc.find(lower_keyword) != std::string::npos) return true;
    }
    
    // 检查标签
    for (const auto& tag : tags) {
        std::string lower_tag = tag;
        std::transform(lower_tag.begin(), lower_tag.end(), lower_tag.begin(), ::tolower);
        if (lower_tag.find(lower_keyword) != std::string::npos) return true;
    }
    
    return false;
}

bool Task::has_status(Status s) const {
    return status == s;
}

bool Task::has_tag(const std::string& tag) const {
    return std::find(tags.begin(), tags.end(), tag) != tags.end();
}

std::string Task::priority_to_string() const {
    switch (priority) {
        case Priority::Low: return "low";
        case Priority::Medium: return "medium";
        case Priority::High: return "high";
        case Priority::Urgent: return "urgent";
    }
    return "medium";
}

std::string Task::status_to_string() const {
    switch (status) {
        case Status::Todo: return "todo";
        case Status::InProgress: return "in_progress";
        case Status::Done: return "done";
        case Status::Cancelled: return "cancelled";
    }
    return "todo";
}

Priority Task::priority_from_string(const std::string& s) {
    if (s == "low") return Priority::Low;
    if (s == "high") return Priority::High;
    if (s == "urgent") return Priority::Urgent;
    return Priority::Medium;
}

Status Task::status_from_string(const std::string& s) {
    if (s == "in_progress" || s == "inprogress") return Status::InProgress;
    if (s == "done") return Status::Done;
    if (s == "cancelled" || s == "cancel") return Status::Cancelled;
    return Status::Todo;
}

json Task::to_json() const {
    json j;
    j["id"] = id;
    j["title"] = title;
    j["priority"] = priority_to_string();
    j["status"] = status_to_string();
    j["user_id"] = user_id;
    j["tags"] = tags;
    if (description.has_value()) {
        j["description"] = *description;
    }
    j["created_at"] = std::chrono::system_clock::to_time_t(created_at);
    j["updated_at"] = std::chrono::system_clock::to_time_t(updated_at);
    return j;
}

Task Task::from_json(const json& j) {
    Task task(j["id"].get<uint32_t>(), j["title"].get<std::string>(), j["user_id"].get<uint32_t>());
    task.priority = priority_from_string(j.value("priority", "medium"));
    task.status = status_from_string(j.value("status", "todo"));
    if (j.contains("description") && !j["description"].is_null()) {
        task.description = j["description"].get<std::string>();
    }
    if (j.contains("tags")) {
        task.tags = j["tags"].get<std::vector<std::string>>();
    }
    if (j.contains("created_at")) {
        task.created_at = std::chrono::system_clock::from_time_t(j["created_at"].get<std::time_t>());
    }
    if (j.contains("updated_at")) {
        task.updated_at = std::chrono::system_clock::from_time_t(j["updated_at"].get<std::time_t>());
    }
    return task;
}

File: TaskCLI/rust/src/model/task.rs
// src/model/task.rs - 任务模型
// 展示 Rust 的结构体、枚举、impl 方法

use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};

/// 任务优先级
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum Priority {
    Low,
    Medium,
    High,
    Urgent,
}

impl Default for Priority {
    fn default() -> Self {
        Priority::Medium
    }
}

impl std::fmt::Display for Priority {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Priority::Low => write!(f, "low"),
            Priority::Medium => write!(f, "medium"),
            Priority::High => write!(f, "high"),
            Priority::Urgent => write!(f, "urgent"),
        }
    }
}

impl std::str::FromStr for Priority {
    type Err = String;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        match s.to_lowercase().as_str() {
            "low" => Ok(Priority::Low),
            "medium" => Ok(Priority::Medium),
            "high" => Ok(Priority::High),
            "urgent" => Ok(Priority::Urgent),
            _ => Err(format!("Unknown priority: {}", s)),
        }
    }
}

/// 任务状态
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum Status {
    Todo,
    InProgress,
    Done,
    Cancelled,
}

impl Default for Status {
    fn default() -> Self {
        Status::Todo
    }
}

impl std::fmt::Display for Status {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Status::Todo => write!(f, "todo"),
            Status::InProgress => write!(f, "in_progress"),
            Status::Done => write!(f, "done"),
            Status::Cancelled => write!(f, "cancelled"),
        }
    }
}

impl std::str::FromStr for Status {
    type Err = String;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        match s.to_lowercase().as_str() {
            "todo" => Ok(Status::Todo),
            "in_progress" | "inprogress" => Ok(Status::InProgress),
            "done" => Ok(Status::Done),
            "cancelled" | "cancel" => Ok(Status::Cancelled),
            _ => Err(format!("Unknown status: {}", s)),
        }
    }
}

/// 任务结构体
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Task {
    pub id: u32,
    pub title: String,
    pub description: Option<String>,
    pub priority: Priority,
    pub status: Status,
    pub tags: Vec<String>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
    pub user_id: u32,
}

impl Task {
    /// 创建新任务（构造函数）
    pub fn new(id: u32, title: String, user_id: u32) -> Self {
        let now = Utc::now();
        Self {
            id,
            title,
            description: None,
            priority: Priority::default(),
            status: Status::default(),
            tags: Vec::new(),
            created_at: now,
            updated_at: now,
            user_id,
        }
    }

    /// 设置描述
    pub fn with_description(mut self, description: String) -> Self {
        self.description = Some(description);
        self
    }

    /// 设置优先级
    pub fn with_priority(mut self, priority: Priority) -> Self {
        self.priority = priority;
        self
    }

    /// 设置标签
    pub fn with_tags(mut self, tags: Vec<String>) -> Self {
        self.tags = tags;
        self
    }

    /// 更新任务状态
    pub fn update_status(&mut self, status: Status) {
        self.status = status;
        self.updated_at = Utc::now();
    }

    /// 更新优先级
    pub fn update_priority(&mut self, priority: Priority) {
        self.priority = priority;
        self.updated_at = Utc::now();
    }

    /// 更新标题
    pub fn update_title(&mut self, title: String) {
        self.title = title;
        self.updated_at = Utc::now();
    }

    /// 检查是否匹配搜索关键词
    pub fn matches(&self, keyword: &str) -> bool {
        let keyword = keyword.to_lowercase();
        self.title.to_lowercase().contains(&keyword)
            || self
                .description
                .as_ref()
                .map(|d| d.to_lowercase().contains(&keyword))
                .unwrap_or(false)
            || self
                .tags
                .iter()
                .any(|t| t.to_lowercase().contains(&keyword))
    }

    /// 检查是否匹配状态
    pub fn has_status(&self, status: Status) -> bool {
        self.status == status
    }

    /// 检查是否包含标签
    pub fn has_tag(&self, tag: &str) -> bool {
        self.tags.iter().any(|t| t == tag)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_task_creation() {
        let task = Task::new(1, "Test Task".to_string(), 1);
        assert_eq!(task.id, 1);
        assert_eq!(task.title, "Test Task");
        assert_eq!(task.status, Status::Todo);
        assert_eq!(task.priority, Priority::Medium);
    }

    #[test]
    fn test_task_builder() {
        let task = Task::new(1, "Test".to_string(), 1)
            .with_description("Description".to_string())
            .with_priority(Priority::High)
            .with_tags(vec!["work".to_string()]);

        assert_eq!(task.description, Some("Description".to_string()));
        assert_eq!(task.priority, Priority::High);
        assert_eq!(task.tags, vec!["work"]);
    }

    #[test]
    fn test_task_matches() {
        let task = Task::new(1, "Buy milk".to_string(), 1)
            .with_description("Get from store")
            .with_tags(vec!["shopping".to_string()]);

        assert!(task.matches("milk"));
        assert!(task.matches("store"));
        assert!(task.matches("shopping"));
        assert!(!task.matches("xyz"));
    }
}
